Long filenames

tordbb opened the pull request:
Fix that truncates filenames during preprocessing, so that the file path is no longer than 255 characters long.

This closes #1

Commit:
Truncating filenames to make sure files are found

Long file names caused files not being found while removing non-images.
Therefore, truncate_filenames() was added to set a maximum filename
length of 40 signs

Commit:
Merge branch 'utils_test' of https://github.com/acntech/fruit-classifier into long_filenames

trying to pull inn Fabian's fixes

Commit:
Merge branch 'master' of https://github.com/acntech/fruit-classifier into long_filenames

Commit:
Changed to max pathlength 255

Max filename is now determined by length of its entire path being 255
characters or less.

Commit:
Started adding test for truncate_filenames

Commit:
Merge branch 'master' of https://github.com/acntech/fruit-classifier into long_filenames

Commit:
added test for truncate_filenames

Commit:
Changed variable names to shorten code line widths

loeiten commented on fruit_classifier/preprocessing/preprocessing_utils.py:
```diff
@@ -4,6 +4,45 @@
 from skimage.transform import resize
 from pathlib import Path
 from fruit_classifier.utils.file_utils import copytree
+import os
```
Use `pathlib`

loeiten commented on fruit_classifier/preprocessing/preprocessing_utils.py:
```diff
@@ -4,6 +4,45 @@
+
+
+def truncate_filenames(raw_dir):
+    # Assumes raw_dir is a directory that contains ONLY directories,
```
Remember docstring of numpy type

loeiten commented on fruit_classifier/preprocessing/preprocessing_utils.py:
```diff
@@ -4,6 +4,45 @@
+    # in which all files/folders will be truncated if their total
+    # path length is more than 255 characters.
+
+    folder_list = os.listdir(raw_dir)
```
Recall that the name `folder` is Windows specific. Use `directory` or `dir` to make it multi purpose

```suggestion
    dir_list = Path(raw_dir).glob('*')
```

loeiten commented on test/preprocessing/test_preprocessing.py:
```diff
@@ -42,6 +46,50 @@ def test_verify_pixel_variation(self):
         self.assertGreater(np.amax(self.comp), np.amin(self.comp))
 
+    def test_truncate_filenames(self):
+        '''
```
```suggestion
        """
```

loeiten commented on test/preprocessing/test_preprocessing.py:
```diff
@@ -42,6 +46,50 @@ def test_verify_pixel_variation(self):
+        run truncate_filenames, and check that the filename now is
+        sufficiently short. However, it was hard to create a file wih
+        long filename, so I did not spend more time on this.
+        '''
```
```suggestion
        """
```

loeiten commented on test/preprocessing/test_preprocessing.py:
```diff
@@ -42,6 +46,50 @@ def test_verify_pixel_variation(self):
 
+    def test_truncate_filenames(self):
+        '''
+        This test creates a folder structure with an image in it,
```
Numpy convention follows git convention

```
"""
A short one-liner

Followed by the longer description
"""
```

loeiten commented on test/preprocessing/test_preprocessing.py:
```diff
@@ -2,6 +2,10 @@
 from pathlib import Path
 import numpy as np
+import shutil
+import os
```
Use pathlib

Commit:
Replaced os in preprocessing_utils

Commit:
Fixed docstrings and replaced os

loeiten commented on fruit_classifier/preprocessing/preprocessing_utils.py:
```diff
@@ -6,6 +6,53 @@
 
+def truncate_filenames(raw_dir):
+    """
+
```
Missing one-liner

loeiten commented on fruit_classifier/preprocessing/preprocessing_utils.py:
```diff
@@ -6,6 +6,53 @@
+    ----------
+    raw_dir: Will truncate files in raw_dir's subfolders
+
+      # Assumes raw_dir is a directory that contains ONLY directories,
```
This should be moved under the one-liner (after a blank line) and without the `#`

loeiten commented on test/preprocessing/test_preprocessing.py:
```diff
@@ -42,6 +45,51 @@ def test_verify_pixel_variation(self):
+        Ideally, this test would create a file with too long filename,
+        run truncate_filenames, and check that the filename now is
+        sufficiently short. However, it was hard to create a file wih
+        long filename, so I did not spend more time on this.
```
This explanation should not reside in the doctring, but rather as an issue (with the label enhancement)

loeiten replied:
I think the `f"{'a'*255}.jpg"` solution will fix this though

loeiten commented on test/preprocessing/test_preprocessing.py:
```diff
@@ -42,6 +45,51 @@ def test_verify_pixel_variation(self):
+        # Select a unique directory name for a new directory
+        outer_dir_name = 'temp_dir'
+        outer_dir_path = self.test_dir.joinpath(outer_dir_name)
+        while outer_dir_path.is_dir():
```
This directory is not created, so this will never exist. I.e. the `while` statement will always be `False`

tordbb replied:
The test now creates a duplicate folder which forces this while loop to be entered during test.

tordbb replied:
Refactored this function, and added creation of a duplicate folder, so that this while loop indeed is being entered during test.

loeiten commented on test/preprocessing/test_preprocessing.py:
```diff
@@ -42,6 +45,51 @@ def test_verify_pixel_variation(self):
+            outer_dir_path = self.test_dir.joinpath(outer_dir_name)
+        outer_dir_path.mkdir(parents=True, exist_ok=True)
+        inner_dir_path = outer_dir_path.joinpath('inner_dir')
+        inner_dir_path.mkdir(parents=True, exist_ok=True)
```
As `parents` is `True`, it will generate `outer_dir_path` if it does not already exist

tordbb replied:
That is the intention. Can you explain the problem?

tordbb replied:
No problem(?)

loeiten commented on test/preprocessing/test_preprocessing.py:
```diff
@@ -42,6 +45,51 @@ def test_verify_pixel_variation(self):
+
+        # Copy an image into this directory
+        short_image_dest_filename = inner_dir_path.\
+            joinpath("short_image_name.jpg")
```
In order to test that `truncate_filenames` work properly, you could use
```suggestion
            joinpath(f"{'a'*255}.jpg")
```

tordbb replied:
Tried this, but it did not work.

tordbb replied:
Did not work.

loeiten reviewed (changes requested):
Almost there :D

Commit:
Refactored uniquefolder. Creates duplic dir.

Creates duplicated directory for test to actually make a step in the
while loop and generate a longer filename. Also, refactored the while
loop into a function

loeiten committed:
Removed debugging code

loeiten commented on test/preprocessing/test_preprocessing.py:
```diff
@@ -42,6 +45,78 @@ def test_verify_pixel_variation(self):
+        shutil.rmtree(outer_dir_path)
+        shutil.rmtree(duplicate_dir_path)
+
+    def create_new_directory(self, outer_dir_name):
```
Can be removed
